File: BOJ/완전탐색/2529/2529.py
# 2529 부등호
# https://www.acmicpc.net/problem/2529
import sys
input = sys.stdin.readline

maxRes = "-9876543210"
minRes = "9876543210"


def DFS(num, depth):
    global maxRes, minRes
    if depth == k+1:
        maxRes = num if int(num) > int(maxRes) else maxRes
        minRes = num if int(num) < int(minRes) else minRes
        return

    for i in range(10):
        if visited[i]:
            continue

        if depth == 0:
            visited[i] = True
            DFS(num+str(i), depth+1)
            visited[i] = False
        else:
            if A[depth - 1] == "<" and int(num[depth-1]) < i:
                visited[i] = True
                DFS(num+str(i), depth+1)
                visited[i] = False

            if A[depth - 1] == ">" and int(num[depth-1]) > i:
                visited[i] = True
                DFS(num+str(i), depth+1)
                visited[i] = False


k = int(input())
A = list(input().split())
visited = [False] * 10

DFS("", 0)

print(maxRes)
print(minRes)

File: BOJ/완전탐색/2529/2529.cpp
// 2529 부등호
// https://www.acmicpc.net/problem/2529
#include <iostream>
#include <vector>
#include <string>

using namespace std;

string res1 = "-9876543210", res2 = "9876543210";
int K;
vector<char> v;
vector<int> visited(10, 0);

void DFS(string currentNumber, int depth) {
    if (depth == K+1)
    {

        if (stoll(res1) < stoll(currentNumber)) res1 = currentNumber;
        if (stoll(res2) > stoll(currentNumber)) res2 = currentNumber;

        return;
    }

    for (int i = 0; i < 10; ++i)
    {
        if (visited[i]) continue;

        if (depth > 0) 
        {
            if (v[depth-1] == '>' && currentNumber[depth-1] - '0' > i)
            {
                visited[i] = 1;
                DFS(currentNumber+to_string(i), depth+1);
                visited[i] = 0;
            }

            if (v[depth-1] == '<' && currentNumber[depth-1] - '0' < i)
            {
                visited[i] = 1;
                DFS(currentNumber+to_string(i), depth+1);
                visited[i] = 0;
            }
        } else
        {
            visited[i] = 1;
            DFS(currentNumber+to_string(i), depth+1);
            visited[i] = 0;
        }
    }

    return;
}

void solution(void) {

    DFS("" ,0);

    cout << res1 << '\n' << res2 << '\n';
    return;
}

int main(void) {
    cin >> K;
    v = vector<char>(K);

    for (int i = 0; i < K; ++i) 
    {
        cin >> v[i];
    }

    solution();
    return 0;
}